crypto/fipsmodule/ml_kem/ml_kem_ref/kem.c

```diff
@@ -59,6 +59,90 @@ int crypto_kem_keypair(ml_kem_params *params,
 
+// FIPS 203. Algorithm 3 BitsToBytes
+// Converts a bit array (of a length that is a multiple of eight)
+// into an array of bytes.
+static void bits_to_bytes(uint8_t *bytes, size_t num_bytes,
+                          const uint8_t *bits, size_t num_bits) {
+  assert(num_bits == num_bytes * 8);
+
+  for (size_t i = 0; i < num_bytes; i++) {
+    uint8_t byte = 0;
+    for (size_t j = 0; j < 8; j++) {
+      byte |= (bits[i * 8 + j] << j);
+    }
+    bytes[i] = byte;
+  }
+}
+
+// FIPS 203. Algorithm 4 BytesToBits
+// Performs the inverse of BitsToBytes, converting a byte array into a bit array.
+static void bytes_to_bits(uint8_t *bits, size_t num_bits,
+                          const uint8_t *bytes, size_t num_bytes) {
+  assert(num_bits == num_bytes * 8);
+
+  for (size_t i = 0; i < num_bytes; i++) {
+    uint8_t byte = bytes[i];
+    for (size_t j = 0; j < 8; j++) {
+      bits[i * 8 + j] = (byte >> j) & 1;
+    }
+  }
+}
+
+#define BYTE_ENCODE_12_IN_SIZE  (256)
+#define BYTE_ENCODE_12_OUT_SIZE (32 * 12)
+#define BYTE_ENCODE_12_NUM_BITS (256 * 12)
+
+// FIPS 203. Algorithm 5 ByteEncode_12
+// Encodes an array of 256 12-bit integers into a byte array.
+static void byte_encode_12(uint8_t out[BYTE_ENCODE_12_OUT_SIZE],
+                           const int16_t in[BYTE_ENCODE_12_IN_SIZE]) {
+  uint8_t bits[BYTE_ENCODE_12_NUM_BITS] = {0};
+  for (size_t i = 0; i < BYTE_ENCODE_12_IN_SIZE; i++) {
+    int16_t a = in[i];
+    for (size_t j = 0; j < 12; j++) {
+      bits[i * 12 + j] = a & 1;
+      a = a >> 1;
+    }
+  }
+  bits_to_bytes(out, BYTE_ENCODE_12_OUT_SIZE, bits, BYTE_ENCODE_12_NUM_BITS);
+}
+
+#define BYTE_DECODE_12_OUT_SIZE (256)
+#define BYTE_DECODE_12_IN_SIZE  (32 * 12)
+#define BYTE_DECODE_12_NUM_BITS (256 * 12)
+
+// FIPS 203. Algorithm 5 ByteDecode_12
+// Decodes a byte array into an array of 256 12-bit integers.
+static void byte_decode_12(int16_t out[BYTE_DECODE_12_OUT_SIZE],
+                           const uint8_t in[BYTE_DECODE_12_IN_SIZE]) {
+  uint8_t bits[BYTE_DECODE_12_NUM_BITS] = {0};
+  bytes_to_bits(bits, BYTE_DECODE_12_NUM_BITS, in, BYTE_DECODE_12_IN_SIZE);
+  for (size_t i = 0; i < BYTE_DECODE_12_OUT_SIZE; i++) {
+    int16_t val = 0;
+    for (size_t j = 0; j < 12; j++) {
+      val |= bits[i * 12 + j] << j;
+    }
+    out[i] = val % KYBER_Q; // TODO(dkostic): make CT.
+  }
+}
+
+#define ENCAPS_KEY_ENCODED_MAX_SIZE (BYTE_ENCODE_12_OUT_SIZE * KYBER_K_MAX)
+#define ENCAPS_KEY_DECODED_MAX_SIZE (BYTE_DECODE_12_OUT_SIZE * KYBER_K_MAX)
+
+// FIPS 203. Section 7.2 Encapsulation key check.
+static int encapsulation_key_modulus_check(ml_kem_params *params, const uint8_t *ek) {
+
+  int16_t ek_decoded[ENCAPS_KEY_DECODED_MAX_SIZE];
+  uint8_t ek_recoded[ENCAPS_KEY_ENCODED_MAX_SIZE];
+
+  for (size_t i = 0; i < params->k; i++) {
+    byte_decode_12(&ek_decoded[i * BYTE_DECODE_12_OUT_SIZE], &ek[i * BYTE_DECODE_12_IN_SIZE]);
+    byte_encode_12(&ek_recoded[i * BYTE_ENCODE_12_OUT_SIZE], &ek_decoded[i * BYTE_ENCODE_12_IN_SIZE]);
```

Comment:
suggest setting `#define BYTE_ENCODE_12_OUT_SIZE BYTE_DECODE_12_IN_SIZE` above. if those values get out of sync, these encode/decode operations would become misaligned.

Reply:
Similarly, the `DECODE_12` macros can be defined in terms of `ENCODE_12` macros.

Reply:
I did that at first but it came out confusing (to me). The definitions as they are now are easy to map to the algorithm definitions in the FIPS spec so I think I'd keep it as is if you don't mind.

Reply:
that's reasonable